akka-persistence-query/src/main/scala/akka/persistence/query/EventEnvelope.scala

```diff
@@ -4,8 +4,76 @@
 
 package akka.persistence.query
 
+import scala.runtime.AbstractFunction4
+
+import akka.util.HashCode
+
+// for binary compatibility (used to be a case class)
+object EventEnvelope extends AbstractFunction4[Offset, String, Long, Any, EventEnvelope] {
+  def apply(offset: Offset, persistenceId: String, sequenceNr: Long, event: Any, timestamp: Long): EventEnvelope =
+    new EventEnvelope(offset, persistenceId, sequenceNr, event, timestamp)
+
+  override def apply(offset: Offset, persistenceId: String, sequenceNr: Long, event: Any): EventEnvelope =
+    new EventEnvelope(offset, persistenceId, sequenceNr, event)
+
+  def unapply(arg: EventEnvelope): Option[(Offset, String, Long, Any)] =
+    Some((arg.offset, arg.persistenceId, arg.sequenceNr, arg.timestamp))
+
+}
+
 /**
  * Event wrapper adding meta data for the events in the result stream of
  * [[akka.persistence.query.scaladsl.EventsByTagQuery]] query, or similar queries.
+ *
+ * The `timestamp` is the time the event was stored, in milliseconds since midnight, January 1, 1970 UTC
+ * (same as `System.currentTimeMillis`).
  */
-final case class EventEnvelope(offset: Offset, persistenceId: String, sequenceNr: Long, event: Any)
+final class EventEnvelope(
+    val offset: Offset,
+    val persistenceId: String,
+    val sequenceNr: Long,
+    val event: Any,
+    val timestamp: Long)
```

Comment:
the reason for using `Long` instead of some more rich type such as `java.time.Instant` is that `Long` has less performance overhead, and millis since epoch is a well established way of representing timestamps.

Reply:
I would only use Long for that reason, good. Then anything wanting to read it with another translation is free to do so.